### PR TITLE
add new apps controllers and support subscription manifest

### PR DESCRIPTION
**Pull request by qmhu**
#### What type of PR is this?
kind/feature

#### What this PR does / why we need it:
support subscription to refer a manifest and create a base object recording to the manifest

#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: pkg/controllers/apps/subscription/subscription.go [hunk 1]
```diff
@@ -19,6 +19,7 @@ package subscription
 import (
 	"context"
 	"fmt"
+	corev1 "k8s.io/api/core/v1"
```
Comment (dixudx) [inline pkg/controllers/apps/subscription/subscription.go:22]: format the imports

File: pkg/controllers/apps/subscription/subscription.go [hunk 2]
```diff
@@ -136,10 +141,11 @@ func (c *Controller) addSubscription(obj interface{}) {
 		if err == nil {
 			msg := fmt.Sprintf("successfully inject finalizer %s to Subscription %s", known.AppFinalizer, klog.KObj(subs))
 			klog.V(4).Info(msg)
-			// todo: add recorder
+			c.recorder.Event(subs, corev1.EventTypeNormal, "FinalizerInjected", msg)
```
Comment (dixudx) [inline pkg/controllers/apps/subscription/subscription.go:144]: here the finalizer is successfully injected

**Review by dixudx — CHANGES_REQUESTED** [review]
please rebase with master

File: pkg/apis/apps/v1alpha1/subscription.go [hunk 3]
```diff
@@ -115,6 +118,25 @@ type Feed struct {
 	FeedSelector *metav1.LabelSelector `json:"feedSelector,omitempty"`
 }
 
+func (f *Feed) Matches(manifest *Manifest) bool {
```
Comment (dixudx) [inline pkg/apis/apps/v1alpha1/subscription.go:121]: better not declare this kind of function in api types

**Review by dixudx — CHANGES_REQUESTED** [review]

File: pkg/controllers/apps/subscription/subscription.go [hunk 4]
```diff
@@ -136,10 +141,11 @@ func (c *Controller) addSubscription(obj interface{}) {
...
 		} else {
-			klog.WarningDepth(4,
-				fmt.Sprintf("failed to inject finalizer %s to Subscription %s: %v", known.AppFinalizer, klog.KObj(subs), err))
+			msg := fmt.Sprintf("failed to inject finalizer %s to Subscription %s: %v", known.AppFinalizer, klog.KObj(subs), err)
+			klog.WarningDepth(4, msg)
+			c.recorder.Event(subs, corev1.EventTypeWarning, "FailedInjectFinalizer", msg)
```
Comment (dixudx) [inline pkg/controllers/apps/subscription/subscription.go:148]: s/FailedInjectFinalizer/FailedInjectingFinalizer

File: pkg/hub/deployer/deployer.go [hunk 5]
```diff
@@ -172,60 +218,107 @@ func (deployer *Deployer) handleSubscription(subs *appsapi.Subscription) error {
...
+				deployer.recorder.Event(subs, corev1.EventTypeWarning, "FailedRetrievingHelmCharts", msg)
+				return err
+			}
+			charts = append(charts, chartList...)
+		} else {
+			manifestList, err := deployer.getManifestsBySelector(subs, feed)
+			if errors.IsNotFound(err) {
+				msg := fmt.Sprintf("Subscription %s is using a nonexistent k8s Objects, feed %s/%s %s/%s %s", klog.KObj(subs), feed.APIVersion, feed.Kind, feed.Namespace, feed.Name, feed.FeedSelector.String())
+				klog.Error(msg)
+				deployer.recorder.Event(subs, corev1.EventTypeWarning, "NonexistentK8SObjects", msg)
```
Comment (dixudx) [inline pkg/hub/deployer/deployer.go:243]: s/NonexistentK8SObjects/NonexistentObject

File: pkg/hub/deployer/deployer.go [hunk 6]
```diff
@@ -172,60 +218,107 @@ func (deployer *Deployer) handleSubscription(subs *appsapi.Subscription) error {
...
+				msg := fmt.Sprintf("failed to get charts matching %q for Subscription %s: %v", feed, klog.KObj(subs), err)
+				klog.Error(msg)
+				deployer.recorder.Event(subs, corev1.EventTypeWarning, "FailedRetrievingHelmCharts", msg)
+				return err
+			}
+			charts = append(charts, chartList...)
+		} else {
+			manifestList, err := deployer.getManifestsBySelector(subs, feed)
+			if errors.IsNotFound(err) {
+				msg := fmt.Sprintf("Subscription %s is using a nonexistent k8s Objects, feed %s/%s %s/%s %s", klog.KObj(subs), feed.APIVersion, feed.Kind, feed.Namespace, feed.Name, feed.FeedSelector.String())
```
Comment (dixudx) [inline pkg/hub/deployer/deployer.go:241]:
```go
fmt.Sprintf("Subscription %s is using a nonexistent Object (APIVersion: %s, Kind: %s with Name %s in Namespace %s that matches %q)", klog.KObj(subs), feed.APIVersion, feed.Kind, feed.Name, feed.Namespace, feed.FeedSelector.String())
```

File: pkg/hub/deployer/deployer.go [hunk 7]
```diff
@@ -172,60 +218,107 @@ func (deployer *Deployer) handleSubscription(subs *appsapi.Subscription) error {
...
+				klog.Error(msg)
+				deployer.recorder.Event(subs, corev1.EventTypeWarning, "NonexistentK8SObjects", msg)
+				return nil
+			}
+			if err != nil {
+				msg := fmt.Sprintf("failed to get k8s Objects matching %q for Subscription %s: %v", feed, klog.KObj(subs), err)
```
Comment (dixudx) [inline pkg/hub/deployer/deployer.go:247]: s/k8s Objects/Object

File: pkg/hub/deployer/deployer.go (hunk 6 above)
Comment (dixudx) [reply pkg/hub/deployer/deployer.go:241]: or we have a function to format `Feed` instead.

File: pkg/hub/deployer/deployer.go [hunk 8]
```diff
@@ -172,60 +218,107 @@ func (deployer *Deployer) handleSubscription(subs *appsapi.Subscription) error {
...
+				klog.Error(msg)
+				deployer.recorder.Event(subs, corev1.EventTypeWarning, "FailedRetrievingK8SObject", msg)
```
Comment (dixudx) [inline pkg/hub/deployer/deployer.go:249]: s/FailedRetrievingK8SObject/FailedRetrievingObject

File: pkg/hub/deployer/deployer.go [hunk 9]
```diff
@@ -172,60 +218,107 @@ func (deployer *Deployer) handleSubscription(subs *appsapi.Subscription) error {
...
+			}
+			manifests = append(manifests, manifestList...)
 		}
-		charts = append(charts, chartList...)
 	}
 
-	if len(charts) == 0 {
-		deployer.recorder.Event(subs, corev1.EventTypeWarning, "NoHelmCharts", "No helm charts get matched")
+	if len(charts) == 0 && len(manifests) == 0 {
+		deployer.recorder.Event(subs, corev1.EventTypeWarning, "NoMatchedFeeds", "No feeds get matched")
```
Comment (dixudx) [inline pkg/hub/deployer/deployer.go:257]: s/NoMatchedFeeds/NoFeedsMatched

File: pkg/hub/deployer/deployer.go [hunk 10]
```diff
@@ -172,60 +218,107 @@ func (deployer *Deployer) handleSubscription(subs *appsapi.Subscription) error {
...
-	// verify HelmChart can be found
-	var chartRefs []appsapi.ChartReference
-	for _, chart := range charts {
-		if len(chart.Status.Phase) == 0 {
-			msg := fmt.Sprintf("HelmChart %s is in verifying", klog.KObj(chart))
-			klog.Warning(msg)
-			deployer.recorder.Event(subs, corev1.EventTypeWarning, "VerifyingHelmChart", msg)
-			return fmt.Errorf(msg)
-		}
+	var allErrs []error
```
Comment (dixudx) [inline pkg/hub/deployer/deployer.go:261]: Should populate to `Base` object here.

And in `handleBase`, we will populate `Description`.

File: pkg/hub/deployer/deployer.go [hunk 11]
```diff
@@ -172,60 +218,107 @@ func (deployer *Deployer) handleSubscription(subs *appsapi.Subscription) error {
...
+		deployer.recorder.Event(subs, corev1.EventTypeNormal, "ManifestsMatched", "manifests get matched")
+		allErrs = append(allErrs, deployer.populateBasesForK8SObjects(subs, manifests)...)
+	}
 
-	return deployer.populateDescriptionsForHelm(subs, chartRefs)
+	return utilerrors.NewAggregate(allErrs)
+}
+
+func (deployer *Deployer) handleManifest(manifest *appsapi.Manifest) error {
+	// TODO: handle deletion
```
Comment (dixudx) [inline pkg/hub/deployer/deployer.go:298]: Please implement it.

File: pkg/hub/deployer/deployer.go [hunk 12]
```diff
@@ -417,3 +567,225 @@ func (deployer *Deployer) handleDescription(desc *appsapi.Description) error {
 
 	return nil
 }
+
+func (deployer *Deployer) populateBasesForK8SObjects(subs *appsapi.Subscription, manifests []*appsapi.Manifest) (errs []error) {
```
Comment (dixudx) [inline pkg/hub/deployer/deployer.go:571]: s/populateBasesForK8SObjects/populateBases

File: pkg/hub/deployer/deployer.go [hunk 13]
```diff
@@ -172,60 +218,107 @@ func (deployer *Deployer) handleSubscription(subs *appsapi.Subscription) error {
...
+
+func (deployer *Deployer) handleManifest(manifest *appsapi.Manifest) error {
+	// TODO: handle deletion
+
+	subscriptions, err := deployer.subsLister.List(labels.NewSelector())
+	if err != nil {
+		return err
+	}
+
+	var matched *appsapi.Subscription
```
Comment (dixudx) [inline pkg/hub/deployer/deployer.go:305]: hold for later review

File: pkg/hub/deployer/deployer.go [hunk 14]
```diff
@@ -417,3 +567,225 @@ func (deployer *Deployer) handleDescription(desc *appsapi.Description) error {
...
+			Name:      subs.Name,
+			Namespace: cluster.Namespace,
+			Labels: map[string]string{
+				known.ObjectCreatedByLabel: known.ClusternetHubName,
+				known.ClusterIDLabel:       cluster.Labels[known.ClusterIDLabel],
+				known.ClusterNameLabel:     cluster.Labels[known.ClusterNameLabel],
+				known.ConfigKindLabel:      subscriptionKind.Kind,
+				known.ConfigNameLabel:      subs.Name,
+				known.ConfigNamespaceLabel: subs.Namespace,
+				known.ConfigUIDLabel:       string(subs.UID),
```
Comment (dixudx) [inline pkg/hub/deployer/deployer.go:724]: append labels that refers to subscription object

File: pkg/hub/deployer/deployer.go [hunk 15]
```diff
@@ -417,3 +567,225 @@ func (deployer *Deployer) handleDescription(desc *appsapi.Description) error {
...
+		Spec: appsapi.BaseSpec{
+			Feeds: feeds,
+		},
+	}
+
+	return base
+}
+
+func (deployer *Deployer) syncBases(subs *appsapi.Subscription, base *appsapi.Base) error {
+	existBase, err := deployer.baseLister.Bases(base.Namespace).Get(base.Name)
```
Comment (dixudx) [inline pkg/hub/deployer/deployer.go:739]: s/existBase/curBase

**Review by dixudx — CHANGES_REQUESTED** [review]

File: pkg/hub/deployer/deployer.go (hunk 14 above)
Comment (dixudx) [reply pkg/hub/deployer/deployer.go:724]: xref #40

File: pkg/hub/deployer/deployer.go (hunk 10 above)
Comment (qmhu) [reply pkg/hub/deployer/deployer.go:261]: will handle HelmChart in another pr later

File: pkg/hub/deployer/deployer.go [hunk 16]
```diff
@@ -162,6 +182,31 @@ func (deployer *Deployer) handleSubscription(subs *appsapi.Subscription) error {
...
+				continue
+			}
+			err = deployer.clusternetclient.AppsV1alpha1().Descriptions(base.Namespace).Delete(context.TODO(), base.Name, metav1.DeleteOptions{
+				PropagationPolicy: &deletePropagationBackground,
+			})
+			if err != nil {
+				return err
+			}
+		}
+		if descs != nil {
```
Comment (dixudx) [inline pkg/hub/deployer/deployer.go:206]: should be `bases` here.

File: pkg/hub/deployer/deployer.go [hunk 17]
```diff
@@ -162,6 +182,31 @@ func (deployer *Deployer) handleSubscription(subs *appsapi.Subscription) error {
...
+		if err != nil {
+			return err
+		}
+
+		// delete all matching Base
+		for _, base := range bases {
+			if base.DeletionTimestamp != nil {
+				continue
+			}
+			err = deployer.clusternetclient.AppsV1alpha1().Descriptions(base.Namespace).Delete(context.TODO(), base.Name, metav1.DeleteOptions{
```
Comment (dixudx) [inline pkg/hub/deployer/deployer.go:199]: should be `Bases(base.Namespace)` here

**Review by dixudx — CHANGES_REQUESTED** [review]

File: pkg/hub/deployer/deployer.go [hunk 18]
```diff
@@ -172,60 +217,115 @@ func (deployer *Deployer) handleSubscription(subs *appsapi.Subscription) error {
...
-		if len(chart.Status.Phase) == 0 {
-			msg := fmt.Sprintf("HelmChart %s is in verifying", klog.KObj(chart))
-			klog.Warning(msg)
-			deployer.recorder.Event(subs, corev1.EventTypeWarning, "VerifyingHelmChart", msg)
-			return fmt.Errorf(msg)
+	var matched *appsapi.Subscription
+	for _, subs := range subscriptions {
+		for _, feed := range subs.Spec.Feeds {
+			if IsFeedMatches(feed, manifest) {
+				matched = subs
```
Comment (dixudx) [inline pkg/hub/deployer/deployer.go:316]: multiple subscriptions could be matched here.

File: pkg/hub/deployer/deployer.go [hunk 19]
```diff
@@ -256,27 +356,84 @@ func (deployer *Deployer) getChartsBySelector(subs *appsapi.Subscription, feed a
...
+	return manifestList, nil
+}
+
+func (deployer *Deployer) populateDescriptionsForHelm(subs *appsapi.Subscription, chartRefs []appsapi.ChartReference) (errs []error) {
 	var mcls []*clusterapi.ManagedCluster
 	for _, subscriber := range subs.Spec.Subscribers {
 		selector, err := metav1.LabelSelectorAsSelector(subscriber.ClusterAffinity)
 		if err != nil {
-			return err
+			errs = append(errs, err)
```
Comment (dixudx) [inline pkg/hub/deployer/deployer.go:419]: it is no use to append error here.

File: pkg/hub/deployer/deployer.go [hunk 20]
```diff
@@ -256,27 +356,84 @@ func (deployer *Deployer) getChartsBySelector(subs *appsapi.Subscription, feed a
...
+			return
 		}
 		clusters, err := deployer.clusterLister.ManagedClusters("").List(selector)
 		if err != nil {
-			return err
+			errs = append(errs, err)
```
Comment (dixudx) [inline pkg/hub/deployer/deployer.go:424]: ditto

File: pkg/hub/deployer/deployer.go [hunk 21]
```diff
@@ -256,27 +356,84 @@ func (deployer *Deployer) getChartsBySelector(subs *appsapi.Subscription, feed a
...
+			return
 		}
 
 		if clusters == nil {
 			deployer.recorder.Event(subs, corev1.EventTypeWarning, "NoClusters", "No clusters get matched")
-			return nil
+			errs = append(errs, err)
```
Comment (dixudx) [inline pkg/hub/deployer/deployer.go:430]: ditto. And here err is nil.

**Review by dixudx — CHANGES_REQUESTED** [review]

File: pkg/hub/deployer/utils.go [hunk 22]
```diff
@@ -0,0 +1,11 @@
+package deployer
+
+import (
+	"fmt"
+
+	appsapi "github.com/clusternet/clusternet/pkg/apis/apps/v1alpha1"
+)
+
+func FormatFeed(feed appsapi.Feed) string {
```
Comment (dixudx) [inline pkg/hub/deployer/utils.go:9]: No need for a new function

File: pkg/hub/deployer/deployer.go (hunk 20 above)
Comment (qmhu) [reply pkg/hub/deployer/deployer.go:424]: refector population for helm and solve these problems

File: pkg/hub/deployer/utils.go (hunk 22 above)
Comment (qmhu) [reply pkg/hub/deployer/utils.go:9]: utils to includes several function including FormatFeed, not only one function now

**Comment by dixudx** [comment]
close.

superceded by #46

**Pull request closed without merging**